// @signalwire/node/dist/common/src/relay/calling/components/Dial.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const BaseComponent_1 = require("./BaseComponent");
const relay_1 = require("../../../util/constants/relay");
const Event_1 = __importDefault(require("../Event"));
class Dial extends BaseComponent_1.BaseComponent {
    constructor() {
        super(...arguments);
        this.eventType = relay_1.CallNotification.State;
        this.method = relay_1.CallMethod.Begin;
        this.controlId = this.call.tag;
    }
    get payload() {
        return {
            tag: this.call.tag,
            device: this.call.device
        };
    }
    notificationHandler(params) {
        this.state = params.call_state;
        const events = [relay_1.CallState.Answered, relay_1.CallState.Ending, relay_1.CallState.Ended];
        this.completed = events.includes(this.state);
        if (this.completed) {
            this.successful = this.state === relay_1.CallState.Answered;
            this.event = new Event_1.default(this.state, params);
        }
        if (this._hasBlocker() && this._eventsToWait.includes(this.state)) {
            this.blocker.resolve();
        }
    }
}
exports.Dial = Dial;
